### PR TITLE
[GLM 5.3 Flash] Fix NaN gradients in chunked KDA

### PR DESCRIPTION
**Pull request by imvladikon**
[CPU CI](https://transformers-ci.lor-e.huggingface.cool/d/pytest-observability-by-pr/pytest-observability-branch?var-pr=48455) [GPU run-slow](https://transformers-ci.lor-e.huggingface.cool/d/pytest-observability-by-pr/pytest-observability-branch?var-pr=48455)

# What does this PR do?
Hi @vasqu thank you for your work, I found an issue (using my dummy 9B model https://huggingface.co/imvladikon/GLM-5.3-Flash-9B-Surgery-Dummy with the same architecture as glm 5.3 flash) and this PR fixes an overflow in that path and also reduces the pairwise decay memory cost. 

### Problem

GLM5.3 Flash uses per-channel KDA decay. After the intra-chunk cumulative sum, `g` has shape `[B, H, N, C, D]`.

The current fallback computes pairwise decay as:

```python
decay_mask = (g.unsqueeze(-2) - g.unsqueeze(-3)).exp()
```

which materializes a `[B, H, N, C, C, D]` tensor.

Each per-token log-decay is non-positive, so cumulative `g` is non-increasing. For causal pairs (`i >= j`), `g_i - g_j <= 0`, but for non-causal pairs (`j > i`) the difference is positive.

With the default lower bound `-5` and `chunk_size=64`, the largest non-causal difference can approach:

```text
5 * (64 - 1) = 315
```

while float32 `exp()` overflows at approximately `88.7`.

The causal mask was applied only after the exponential. The forward output therefore remained finite because the overflowing entries were discarded, but autograd still propagated through the overflowing `exp()`. On those masked entries the backward effectively evaluates `0 * inf`, producing `NaN` gradients.

The pairwise representation is also very expensive in memory. For GLM-5.3-Flash (`64` linear heads, `head_dim=128`, `chunk_size=64`), the decay tensor alone is approximately:

- `0.25 GiB` at sequence length `128`
- `8 GiB` at sequence length `4096`

before accounting for the rest of the activations.

### Fix

Instead of materializing the full per-channel pairwise decay tensor, compute the two causal products required by KDA, `Aqk` and `Akk`, blockwise.

For a causal cross-block where every row comes after every column, choose a cumulative-decay reference `A = g[mid]`. Then:

```text
exp(g_i - g_j)
= exp(g_i - A) * exp(A - g_j)
```

Since cumulative `g` is non-increasing and `j < mid <= i`, both exponent arguments are non-positive:

```text
g_i - A <= 0
A - g_j <= 0
```

The weighted decay product can therefore be evaluated exactly as a matrix multiplication over `head_dim`, without constructing `[rows, columns, head_dim]` and without ever forming a positive cross-block exponent.

This is the same local-reference factorization used by the production FLA KDA kernels.

FLA KDA:

https://github.com/fla-org/flash-linear-attention/blob/35dceaee5408e69a555fec34cb215c93c375dabe/fla/ops/kda/chunk_intra.py

The relevant computation there is equivalent to:

```python
b_gqn = exp2(b_g1 - b_gn1)
b_kgt = transpose(k0 * exp2(b_gn1 - b_g0))

Aqk += dot(q1 * b_gqn, b_kgt)
Akk += dot(k1 * b_gqn, b_kgt)
```

Meta's KDA implementation documents the same decomposition explicitly:

https://github.com/meta-pytorch/attention-gym/blob/6ee58c2a437d99d93533a85b9ccb6a955358cf20/attn_gym/linear/kda/fwd/triton/chunk_kda_fwd_intra_sub_chunk_forloop.py

It describes splitting:

```text
2^(g_i - g_j)
```

into:

```text
2^(g_i - ref) * 2^(ref - g_j)
```

so that the decay-weighted interaction becomes a matrix multiplication.

The Kimi Linear paper also uses secondary chunking for numerical stability in KDA:

https://arxiv.org/abs/2510.26692

The official Kimi Linear repository points to FLA as the open-source KDA kernel implementation:

https://github.com/MoonshotAI/Kimi-Linear

The PyTorch fallback in this PR applies the same factorization recursively over the causal triangle and keeps the direct pairwise formulation only inside small `8x8` diagonal blocks. Inside those diagonal blocks, non-causal values are masked to `-inf` before `exp()`.

The two KDA products, `Aqk` and `Akk`, share the same decay factors and key-side term and are computed together.

### Alternatives considered

Before switching to the blockwise formulation, two smaller fixes were tested:

| Approach | Prototype | Result | Why not used |
|---|---|---|---|
| Clamp pairwise log-decays before `exp()` | [`04de6850`](https://github.com/imvladikon/transformers/commit/04de685004f314a82b1a7379ee8949f2fdbd6615) | Fixes NaN gradients and leaves causal values unchanged under the current `g <= 0` invariant | It relies on the decay-sign invariant rather than causal structure and still materializes the full `[B, H, N, C, C, D]` tensor. |
| Mask non-causal entries to `-inf` before `exp()` | [`20801e01`](https://github.com/imvladikon/transformers/commit/20801e01fd4f224117678d2fe161bb88116aec1e) | Fixes the overflow exactly at the entries that are discarded; forward output is unchanged and gradients are finite | This is the clean minimal numerical fix, but it still keeps the full per-channel pairwise decay tensor and therefore does not solve the memory problem. |
| Blockwise local-reference factorization | [`e61f4bc2`](https://github.com/imvladikon/transformers/commit/e61f4bc2cbf904cf60da9047cf8d542352bdd58c) | Fixes both numerical overflow and pairwise-decay memory usage | Selected. It computes the same causal `Aqk`/`Akk` products using the local-reference factorization already used by FLA KDA, while only using direct pairwise evaluation inside small diagonal blocks. |

The mask-before-`exp()` version is sufficient if the only objective is to prevent NaN gradients. The blockwise version is used because the same pairwise representation causing the numerical issue is also the dominant memory cost of the PyTorch fallback.

### Related Transformers implementation

Qwen3-Next avoids the numerical part of the same problem by masking non-causal cumulative-decay differences before exponentiation:

https://github.com/huggingface/transformers/blob/main/src/transformers/models/qwen3_next/modeling_qwen3_next.py

Its decay is scalar per head, so the pairwise tensor is only `[B, H, N, C, C]`. GLM5-Next has a separate decay for each head channel, which introduces the additional `D` dimension and makes the direct pairwise representation substantially more expensive.

### Validation

The implementation was compared against both:

1. the pairwise implementation with non-causal entries masked before `exp()`;
2. the recurrent KDA reference.

Across float32 and bfloat16 tests, including partial chunks, different sequence/chunk sizes, `D != V`, L2-normalized Q/K, initial recurrent states, and extreme negative decay values:

- outputs remained finite;
- recurrent states remained finite;
- all gradients remained finite;
- float32 output/state differences were on the order of `1e-7`;
- gradient differences were on the order of `1e-8`;
- bfloat16 differences were within expected matmul precision.

The original implementation reproduces the failure mode directly: the forward result remains finite while gradients become NaN because non-causal pairwise exponentials overflow before masking.

A100, bfloat16, `H=64`, `head_dim=128`, `chunk_size=64`, full forward + backward:

| Sequence length | Implementation | Saved autograd memory | Peak memory | Forward + backward |
|---:|---|---:|---:|---:|
| 512 | pairwise | 3.39 GiB | 6.47 GiB | 48.2 ms |
| 512 | blockwise (`block_size=8`) | 0.88 GiB | 1.88 GiB | 52.0 ms |
| 2048 | pairwise | 13.54 GiB | 30.05 GiB | 182.9 ms |
| 2048 | blockwise (`block_size=8`) | 3.51 GiB | 8.60 GiB | 80.2 ms |

`block_size=8` was selected from a sweep over `4`, `8`, `16`, and `32` on the GLM-5.3-Flash attention shapes. It gave the best overall memory/runtime tradeoff in these tests.

### CODE REVIEW

**Comment by vasqu** [comment]
Hmm, we recently refactored GDN in #47625, would a similar refactor with KDA already solve the overflow? (note that KDA is essentially GDN with and additional dim for the heads)

**Comment by imvladikon** [comment]
@vasqu you mean applying the masking part? Yes, it solves nan gradient issue (I actually tested exactly that as an intermediate fix (20801e01): mask the non-causal g_i - g_j values to -inf before exp().
But there is a memory tradeoff because the remaining issue is that KDA has per-channel decay, as you wrote, so the analogous pairwise_decay has shape [B, H, N, C, C, D], rather than [B, H, N, C, C] as in GDN. We also cannot compute QK^T first and multiply by the decay afterwards, since the decay varies along D.
For GLM 5.3 Flash (H=64, D=128, C=64), that tensor alone is ~8 GiB at seq_len=4096. The blockwise factorization in this PR is intended to solve that second problem as well.

**Comment by vasqu** [comment]
I'd love to go this step by step first, the memory issue would also apply to GDN then so I want to first make the NaN issue then memory issue

Overall, we want users to mostly use hf kernels instead tbh (because the eager path will always lack in comparison)

**Comment by imvladikon** [comment]
@vasqu Makes sense 👌 I’ve simplified the PR to just the NaN fix, I completely agree about preferring the kernels path. I think I happened to hit the fallback case because of the FLA resolution issue fixed in #48221, so I ended up observing the eager-path memory behavior more directly.

File: src/transformers/models/glm5_next/modular_glm5_next.py [hunk 1]
```diff
@@ -542,7 +542,8 @@ def chunk_kimi_delta_attention(
     # Main difference to GDN is the per head application of `g` which was broadcasted across heads instead
     g = g.cumsum(dim=-2)
     mask = torch.triu(torch.ones(chunk_size, chunk_size, dtype=torch.bool, device=query.device), diagonal=0)
-    decay_mask = (g.unsqueeze(-2) - g.unsqueeze(-3)).exp().float()
+    strict_mask = torch.triu(torch.ones(chunk_size, chunk_size, dtype=torch.bool, device=query.device), diagonal=1)
+    decay_mask = (g.unsqueeze(-2) - g.unsqueeze(-3)).masked_fill(strict_mask[..., None], float("-inf")).exp().float()
```
Comment (vasqu) [inline src/transformers/models/glm5_next/modular_glm5_next.py:546]: Can we add a test for glm that could encounter this?

Comment (imvladikon) [reply src/transformers/models/glm5_next/modular_glm5_next.py:546]: @vasqu yes, sure, I added 2 tests in tests/models/glm5_next/test_modeling_glm5_next.py

`test_kda_chunked_large_decay_matches_recurrent` compares chunked vs recurrent KDA (output, final state, all five input gradients; seq 64 and 72) with the gate saturated at linear_lower_bound: before the fix dq, dk, dg, dbeta are NaN while the forward stays finite, after it everything matches within rtol=1e-4, atol=1e-5.

`test_linear_attention_backward_with_saturated_forget_gate`  backward through a real Glm5NextTextLinearAttention with the gate pinned to linear_lower_bound * sigmoid(8): before the fix 8 of 13 parameters plus the input get NaN gradients, after it all are finite.


PS. I might add such test about selected lines:
```python
    def test_kda_decay_mask_matches_the_unmasked_one_on_causal_pairs(self):
        g = -torch.rand(batch_size, num_heads, num_chunks, chunk_size, head_dim, device=torch_device)
        g = g.cumsum(dim=-2)
        strict_mask = torch.triu(torch.ones(chunk_size, chunk_size, dtype=torch.bool, device=torch_device), diagonal=1)
        pairwise_decay = g.unsqueeze(-2) - g.unsqueeze(-3)
        unmasked = pairwise_decay.exp().float()
        masked = pairwise_decay.masked_fill(strict_mask[..., None], float("-inf")).exp().float()
        non_causal = strict_mask[..., None].expand_as(masked)
        torch.testing.assert_close(masked[~non_causal], unmasked[~non_causal], rtol=0, atol=0)
        self.assertTrue((masked[non_causal] == 0).all())
        self.assertTrue(
            torch.isfinite(unmasked).all(), "the reference is only valid while the unmasked decay is finite"
        )
```
but I'm not sure because it doesn't call any production transformers code.

Comment (vasqu) [reply src/transformers/models/glm5_next/modular_glm5_next.py:546]: Hmm not I would avoid not using our code. It's good for internal debugging but not production code

File: tests/models/glm5_next/test_modeling_glm5_next.py [hunk 2]
```diff
@@ -582,6 +587,88 @@ def test_generate_compile_model_forward_fullgraph(self):
...
 
+    def test_kda_chunked_large_decay_matches_recurrent(self):
+        """
+        Ensures that the chunked KDA implementation matches the recurrent one, gradients included, when the forget
+        gate sits at its lower bound. The pairwise decay `g_i - g_j` is positive for non-causal pairs and reaches
+        `|linear_lower_bound| * (chunk_size - 1)`, i.e. 315 by default, while fp32 `exp` overflows above ~88.7.
+        Those entries are dropped from the output but not from the backward pass, so only the gradients turn into NaNs.
+        """
+        torch.manual_seed(0)
+        batch_size, num_heads, head_dim, chunk_size = 1, 2, 16, 64
```
Comment (vasqu) [inline tests/models/glm5_next/test_modeling_glm5_next.py]: Could we use our prepare input and config functions for this?

File: tests/models/glm5_next/test_modeling_glm5_next.py [hunk 3]
```diff
@@ -582,6 +587,88 @@ def test_generate_compile_model_forward_fullgraph(self):
...
                 assert_similar_generate_outputs(dynamic_result, compiled_result, atol=atol, rtol=rtol)
 
+    def test_kda_chunked_large_decay_matches_recurrent(self):
+        """
+        Ensures that the chunked KDA implementation matches the recurrent one, gradients included, when the forget
+        gate sits at its lower bound. The pairwise decay `g_i - g_j` is positive for non-causal pairs and reaches
+        `|linear_lower_bound| * (chunk_size - 1)`, i.e. 315 by default, while fp32 `exp` overflows above ~88.7.
+        Those entries are dropped from the output but not from the backward pass, so only the gradients turn into NaNs.
+        """
+        torch.manual_seed(0)
```
Comment (vasqu) [inline tests/models/glm5_next/test_modeling_glm5_next.py]: seeding needed?

File: tests/models/glm5_next/test_modeling_glm5_next.py [hunk 4]
```diff
@@ -582,6 +587,88 @@ def test_generate_compile_model_forward_fullgraph(self):
...
+        batch_size, num_heads, head_dim, chunk_size = 1, 2, 16, 64
+        lower_bound = Glm5NextTextConfig().linear_lower_bound
+
+        # 64 spans exactly one chunk, 72 spans two chunks with padding
+        for seq_length in [64, 72]:
```
Comment (vasqu) [inline tests/models/glm5_next/test_modeling_glm5_next.py]: can just use one no?

File: tests/models/glm5_next/test_modeling_glm5_next.py [hunk 5]
```diff
@@ -582,6 +587,88 @@ def test_generate_compile_model_forward_fullgraph(self):
...
+
+                out, state = chunk_kimi_delta_attention(*inputs, chunk_size=chunk_size, **kwargs)
+                reference_out, reference_state = recurrent_kimi_delta_attention(*reference_inputs, **kwargs)
+
+                # the same cotangents for both, with the final state in the loss so that its backward is covered too
+                out_cotangent, state_cotangent = torch.randn_like(out), torch.randn_like(state)
+                loss = (out * out_cotangent).sum() + (state * state_cotangent).sum()
+                reference_loss = (reference_out * out_cotangent).sum() + (reference_state * state_cotangent).sum()
+                grads = torch.autograd.grad(loss, inputs)
+                reference_grads = torch.autograd.grad(reference_loss, reference_inputs)
```
Comment (vasqu) [inline tests/models/glm5_next/test_modeling_glm5_next.py]: Could be simpler just on the out no?

File: tests/models/glm5_next/test_modeling_glm5_next.py [hunk 6]
```diff
@@ -582,6 +587,88 @@ def test_generate_compile_model_forward_fullgraph(self):
...
+                    ("output", out, reference_out),
+                    ("state", state, reference_state),
+                    *zip(["dq", "dk", "dv", "dg", "dbeta"], grads, reference_grads),
+                ]:
+                    with self.subTest(tensor=name):
+                        self.assertTrue(torch.isfinite(actual).all(), f"non-finite {name}")
+                        self.assertTrue(torch.isfinite(expected).all(), f"non-finite reference {name}")
+                        torch.testing.assert_close(actual, expected, rtol=1e-4, atol=1e-5)
+
+    def test_linear_attention_backward_with_saturated_forget_gate(self):
```
Comment (vasqu) [inline tests/models/glm5_next/test_modeling_glm5_next.py:587]: Would this test alone not suffice? (including the changes I suggested above)

Comment (vasqu) [reply tests/models/glm5_next/test_modeling_glm5_next.py:587]: We could make sure we run the fallback based on cpu otherwise we might get into fla / kernels territory no?

**Review by vasqu — COMMENTED** [review]
Just about the test design tbh

**Comment by imvladikon** [comment]
@vasqu good points, I rewrote tests(just left only 1 simple test according to your code review comments 👌)

File: tests/models/glm5_next/test_modeling_glm5_next.py [hunk 7]
```diff
@@ -582,6 +584,39 @@ def test_generate_compile_model_forward_fullgraph(self):
...
+        hidden_states = floats_tensor([self.model_tester.batch_size, self.model_tester.seq_length, config.hidden_size])
+        hidden_states = hidden_states.to("cpu").requires_grad_(True)
+
+        output = layer(hidden_states)
+        self.assertTrue(torch.isfinite(output).all())
+
+        output.square().mean().backward()
+        self.assertTrue(torch.isfinite(hidden_states.grad).all(), "non-finite gradient on the inputs")
+        for name, parameter in layer.named_parameters():
+            with self.subTest(parameter=name):
```
Comment (vasqu) [inline tests/models/glm5_next/test_modeling_glm5_next.py]: last nit: let's remove the subtest but then we are good to go 🫡

**Review by vasqu — APPROVED** [review]
See my last comment but then we can merge

**Comment by imvladikon** [comment]
@vasqu done!  🙏 thank you!

File: tests/models/glm5_next/test_modeling_glm5_next.py (hunk 7 above)
Comment (imvladikon) [reply tests/models/glm5_next/test_modeling_glm5_next.py]: @vasqu done! 🙏 thank you!
